// eth-connect/methods/eth.js
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
// tslint:disable:variable-name
import * as formatters from '../utils/formatters';
import * as utils from '../utils/utils';
import { Method } from '../Method';
import { Property } from '../Property';
/**
 * @public
 */
export var eth;
(function (eth) {
    eth.eth_getBalance = new Method({
        callName: 'eth_getBalance',
        params: 2,
        inputFormatter: [formatters.inputAddressFormatter, formatters.inputDefaultBlockNumberFormatter],
        outputFormatter: formatters.outputBigNumberFormatter
    });
    eth.eth_getStorageAt = new Method({
        callName: 'eth_getStorageAt',
        params: 3,
        inputFormatter: [formatters.inputAddressFormatter, utils.toHex, formatters.inputDefaultBlockNumberFormatter],
        outputFormatter: utils.toData
    });
    eth.eth_getCode = new Method({
        callName: 'eth_getCode',
        params: 2,
        inputFormatter: [formatters.inputAddressFormatter, formatters.inputDefaultBlockNumberFormatter],
        outputFormatter: utils.toData
    });
    eth.eth_getBlockByHash = new Method({
        callName: 'eth_getBlockByHash',
        params: 2,
        inputFormatter: [
            formatters.inputBlockNumberFormatter,
            function (val) {
                return !!val;
            }
        ],
        outputFormatter: formatters.outputBlockFormatter
    });
    eth.eth_getBlockByNumber = new Method({
        callName: 'eth_getBlockByNumber',
        params: 2,
        inputFormatter: [
            formatters.inputBlockNumberFormatter,
            function (val) {
                return !!val;
            }
        ],
        outputFormatter: formatters.outputBlockFormatter
    });
    eth.eth_getUncleByBlockHashAndIndex = new Method({
        callName: 'eth_getUncleByBlockHashAndIndex',
        params: 2,
        inputFormatter: [formatters.inputBlockNumberFormatter, utils.toHex],
        outputFormatter: formatters.outputBlockFormatter
    });
    eth.eth_getUncleByBlockNumberAndIndex = new Method({
        callName: 'eth_getUncleByBlockNumberAndIndex',
        params: 2,
        inputFormatter: [formatters.inputBlockNumberFormatter, utils.toHex],
        outputFormatter: formatters.outputBlockFormatter
    });
    eth.eth_getBlockTransactionCountByHash = new Method({
        callName: 'eth_getBlockTransactionCountByHash',
        params: 1,
        inputFormatter: [formatters.inputBlockNumberFormatter],
        outputFormatter: utils.toDecimal
    });
    eth.eth_getBlockTransactionCountByNumber = new Method({
        callName: 'eth_getBlockTransactionCountByNumber',
        params: 1,
        inputFormatter: [formatters.inputBlockNumberFormatter],
        outputFormatter: utils.toDecimal
    });
    eth.eth_getUncleCountByBlockHash = new Method({
        callName: 'eth_getUncleCountByBlockHash',
        params: 1,
        inputFormatter: [formatters.inputBlockNumberFormatter],
        outputFormatter: utils.toDecimal
    });
    eth.eth_getUncleCountByBlockNumber = new Method({
        callName: 'eth_getUncleCountByBlockNumber',
        params: 1,
        inputFormatter: [formatters.inputBlockNumberFormatter],
        outputFormatter: utils.toDecimal
    });
    eth.eth_getTransactionByHash = new Method({
        callName: 'eth_getTransactionByHash',
        params: 1,
        inputFormatter: [formatters.inputTransactionId],
        outputFormatter: formatters.outputTransactionFormatter
    });
    eth.eth_getTransactionByBlockHashAndIndex = new Method({
        callName: 'eth_getTransactionByBlockHashAndIndex',
        params: 2,
        inputFormatter: [formatters.inputBlockNumberFormatter, utils.toHex],
        outputFormatter: formatters.outputTransactionFormatter
    });
    eth.eth_getTransactionByBlockNumberAndIndex = new Method({
        callName: 'eth_getTransactionByBlockNumberAndIndex',
        params: 2,
        inputFormatter: [formatters.inputBlockNumberFormatter, utils.toHex],
        outputFormatter: formatters.outputTransactionFormatter
    });
    eth.eth_getTransactionReceipt = new Method({
        callName: 'eth_getTransactionReceipt',
        params: 1,
        inputFormatter: [formatters.inputTransactionId],
        outputFormatter: formatters.outputTransactionReceiptFormatter
    });
    eth.eth_getTransactionCount = new Method({
        callName: 'eth_getTransactionCount',
        params: 2,
        inputFormatter: [formatters.inputAddressFormatter, formatters.inputDefaultBlockNumberFormatter],
        outputFormatter: utils.toDecimal
    });
    eth.eth_sendRawTransaction = new Method({
        callName: 'eth_sendRawTransaction',
        params: 1,
        inputFormatter: [null /* signed data transaction, hex string */],
        outputFormatter: utils.toData
    });
    eth.web3_sha3 = new Method({
        callName: 'web3_sha3',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toData
    });
    eth.eth_sendTransaction = new Method({
        callName: 'eth_sendTransaction',
        params: 1,
        inputFormatter: [formatters.inputTransactionFormatter],
        outputFormatter: utils.toData
    });
    eth.eth_sign = new Method({
        callName: 'eth_sign',
        params: 2,
        inputFormatter: [formatters.inputAddressFormatter, null],
        outputFormatter: utils.toData
    });
    eth.eth_call = new Method({
        callName: 'eth_call',
        params: 2,
        inputFormatter: [formatters.inputCallFormatter, formatters.inputDefaultBlockNumberFormatter],
        outputFormatter: utils.toData
    });
    eth.eth_estimateGas = new Method({
        callName: 'eth_estimateGas',
        params: 1,
        inputFormatter: [formatters.inputCallFormatter],
        outputFormatter: utils.toDecimal
    });
    eth.eth_submitWork = new Method({
        callName: 'eth_submitWork',
        params: 3,
        inputFormatter: [
            null,
            null,
            null // DATA, 32 Bytes - The mix digest (256 bits)
        ],
        outputFormatter: utils.toBoolean
    });
    eth.eth_getWork = new Method({
        callName: 'eth_getWork',
        params: 0,
        inputFormatter: [],
        outputFormatter: utils.toArray
    });
    eth.eth_coinbase = new Property({
        getter: 'eth_coinbase',
        outputFormatter: utils.toAddress
    });
    eth.eth_mining = new Property({
        getter: 'eth_mining',
        outputFormatter: utils.toBoolean
    });
    eth.eth_hashrate = new Property({
        getter: 'eth_hashrate',
        outputFormatter: utils.toDecimal
    });
    eth.eth_syncing = new Property({
        getter: 'eth_syncing',
        outputFormatter: formatters.outputSyncingFormatter
    });
    eth.eth_gasPrice = new Property({
        getter: 'eth_gasPrice',
        outputFormatter: formatters.outputBigNumberFormatter
    });
    eth.eth_accounts = new Property({
        getter: 'eth_accounts',
        outputFormatter: utils.toArray
    });
    eth.eth_blockNumber = new Property({
        getter: 'eth_blockNumber',
        outputFormatter: utils.toDecimal
    });
    eth.eth_protocolVersion = new Property({
        getter: 'eth_protocolVersion',
        outputFormatter: utils.toDecimal
    });
    eth.web3_clientVersion = new Property({
        getter: 'web3_clientVersion',
        outputFormatter: utils.toString
    });
    eth.net_version = new Property({
        getter: 'net_version',
        outputFormatter: utils.toString
    });
    eth.shh_version = new Method({
        callName: 'shh_version',
        params: 0,
        inputFormatter: [],
        outputFormatter: utils.toDecimal
    });
    // subscribe and unsubscribe missing
    eth.shh_post = new Method({
        callName: 'shh_post',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toBoolean
    });
    eth.personal_newAccount = new Method({
        callName: 'personal_newAccount',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toAddress
    });
    eth.personal_importRawKey = new Method({
        callName: 'personal_importRawKey',
        params: 2,
        inputFormatter: [null /* private key */, null /* passphrase */],
        outputFormatter: utils.toAddress
    });
    eth.personal_sign = new Method({
        callName: 'personal_sign',
        params: 3,
        inputFormatter: [utils.toStringData, formatters.inputAddressFormatter, null],
        outputFormatter: utils.toData
    });
    eth.personal_ecRecover = new Method({
        callName: 'personal_ecRecover',
        params: 2,
        inputFormatter: [utils.toStringData /* message */, null /* signature */],
        outputFormatter: utils.toAddress
    });
    eth.personal_unlockAccount = new Method({
        callName: 'personal_unlockAccount',
        params: 3,
        inputFormatter: [formatters.inputAddressFormatter, null, utils.toNullDecimal],
        outputFormatter: utils.toBoolean
    });
    eth.personal_sendTransaction = new Method({
        callName: 'personal_sendTransaction',
        params: 2,
        inputFormatter: [formatters.inputTransactionFormatter, null],
        outputFormatter: utils.toData
    });
    eth.personal_lockAccount = new Method({
        callName: 'personal_lockAccount',
        params: 1,
        inputFormatter: [formatters.inputAddressFormatter],
        outputFormatter: utils.toBoolean
    });
    eth.personal_listAccounts = new Property({
        getter: 'personal_listAccounts',
        outputFormatter: utils.toArray
    });
    eth.net_listening = new Property({
        getter: 'net_listening',
        outputFormatter: utils.toBoolean
    });
    eth.net_peerCount = new Property({
        getter: 'net_peerCount',
        outputFormatter: utils.toDecimal
    });
    eth.eth_newFilter = new Method({
        callName: 'eth_newFilter',
        params: 1,
        inputFormatter: [formatters.inputFilterOptions],
        outputFormatter: utils.toHex
    });
    eth.eth_getLogs = new Method({
        callName: 'eth_getLogs',
        params: 1,
        inputFormatter: [utils.toHex],
        outputFormatter: utils.toArray
    });
    eth.eth_newBlockFilter = new Method({
        callName: 'eth_newBlockFilter',
        params: 0,
        inputFormatter: [],
        outputFormatter: utils.toHex
    });
    eth.eth_newPendingTransactionFilter = new Method({
        callName: 'eth_newPendingTransactionFilter',
        params: 0,
        inputFormatter: [],
        outputFormatter: utils.toHex
    });
    eth.eth_uninstallFilter = new Method({
        callName: 'eth_uninstallFilter',
        params: 1,
        inputFormatter: [utils.toHex],
        outputFormatter: utils.toBoolean
    });
    eth.eth_getFilterLogs = new Method({
        callName: 'eth_getFilterLogs',
        params: 1,
        inputFormatter: [utils.toHex],
        outputFormatter: utils.toArray
    });
    eth.eth_getFilterChanges = new Method({
        callName: 'eth_getFilterChanges',
        params: 1,
        inputFormatter: [utils.toHex],
        outputFormatter: utils.toArray
    });
    eth.eth_submitHashrate = new Method({
        callName: 'eth_submitHashrate',
        params: 2,
        inputFormatter: [null /* hashrate */, null /* id */],
        outputFormatter: utils.toBoolean
    });
    eth.shh_newIdentity = new Method({
        callName: 'shh_newIdentity',
        params: 0,
        inputFormatter: [],
        outputFormatter: utils.toData
    });
    eth.shh_hasIdentity = new Method({
        callName: 'shh_hasIdentity',
        params: 1,
        inputFormatter: [
            null // DATA, 60 Bytes - The identity address to check.
        ],
        outputFormatter: utils.toBoolean
    });
    eth.shh_newGroup = new Method({
        callName: 'shh_newGroup',
        params: 0,
        inputFormatter: [],
        outputFormatter: utils.toData
    });
    eth.shh_addToGroup = new Method({
        callName: 'shh_addToGroup',
        params: 1,
        inputFormatter: [
            null // DATA, 60 Bytes - The identity address to add to a group (?)
        ],
        outputFormatter: utils.toBoolean
    });
    eth.shh_newFilter = new Method({
        callName: 'shh_newFilter',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toHex
    });
    eth.shh_uninstallFilter = new Method({
        callName: 'shh_uninstallFilter',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toBoolean
    });
    eth.shh_getLogs = new Method({
        callName: 'shh_getLogs',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toArray
    });
    eth.shh_getFilterMessages = new Method({
        callName: 'shh_getFilterMessages',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toArray
    });
    eth.shh_getFilterChanges = new Method({
        callName: 'shh_getFilterChanges',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toArray
    });
    eth.shh_getMessages = new Method({
        callName: 'shh_getMessages',
        params: 1,
        inputFormatter: [null],
        outputFormatter: utils.toArray
    });
})(eth || (eth = {}));
