// @dcl/ecs/dist/systems/raycast.js
import * as components from '../components';
import { EntityState } from '../engine/entity';
/**
 * @internal
 */
export function createRaycastSystem(engine) {
    const raycastComponent = components.Raycast(engine);
    const raycastResultComponent = components.RaycastResult(engine);
    const entitiesCallbackResultMap = new Map();
    const defaultOptions = {
        maxDistance: 16,
        queryType: 0 /* RaycastQueryType.RQT_HIT_FIRST */,
        continuous: false,
        originOffset: { x: 0, y: 0, z: 0 },
        collisionMask: 2 /* ColliderLayer.CL_PHYSICS */
    };
    const getLocalDirectionRaycastDefaultOptions = (options = {}) => ({
        ...defaultOptions,
        ...options,
        directionRawValue: {
            $case: 'localDirection',
            localDirection: options.direction || { x: 0, y: 0, z: 1 }
        }
    });
    const getGlobalDirectionRaycastDefaultOptions = (options = {}) => ({
        ...defaultOptions,
        ...options,
        directionRawValue: {
            $case: 'globalDirection',
            globalDirection: options.direction || { x: 0, y: 0, z: 1 }
        }
    });
    const getGlobalTargetRaycastDefaultOptions = (options = {}) => ({
        ...defaultOptions,
        ...options,
        directionRawValue: {
            $case: 'globalTarget',
            globalTarget: options.target || { x: 0, y: 0, z: 0 }
        }
    });
    const getTargetEntityRaycastDefaultOptions = (options = {}) => ({
        ...defaultOptions,
        ...options,
        directionRawValue: {
            $case: 'targetEntity',
            targetEntity: options.targetEntity || 0
        }
    });
    function registerRaycast(entity, callback, options) {
        const raycast = raycastComponent.getOrCreateMutable(entity);
        raycast.maxDistance = options.maxDistance;
        raycast.originOffset = options.originOffset;
        raycast.collisionMask = options.collisionMask;
        raycast.direction = options.directionRawValue;
        raycast.continuous = options.continuous;
        raycast.queryType = options.queryType;
        entitiesCallbackResultMap.set(entity, { callback: callback });
    }
    function removeRaycast(entity) {
        raycastComponent.deleteFrom(entity);
        raycastResultComponent.deleteFrom(entity);
        entitiesCallbackResultMap.delete(entity);
    }
    // @internal
    engine.addSystem(function EventSystem() {
        for (const [entity, data] of entitiesCallbackResultMap) {
            const raycast = raycastComponent.getOrNull(entity);
            if (engine.getEntityState(entity) === EntityState.Removed || !raycast) {
                entitiesCallbackResultMap.delete(entity);
                continue;
            }
            const currentResult = raycastResultComponent.getOrNull(entity);
            if (!currentResult)
                continue;
            data.callback(currentResult);
            if (!raycast.continuous)
                removeRaycast(entity);
        }
    });
    return {
        removeRaycasterEntity(entity) {
            removeRaycast(entity);
        },
        registerLocalDirectionRaycast(entity, callback, opts) {
            registerRaycast(entity, callback, getLocalDirectionRaycastDefaultOptions(opts));
        },
        registerGlobalDirectionRaycast(entity, callback, opts) {
            registerRaycast(entity, callback, getGlobalDirectionRaycastDefaultOptions(opts));
        },
        registerGlobalTargetRaycast(entity, callback, opts) {
            registerRaycast(entity, callback, getGlobalTargetRaycastDefaultOptions(opts));
        },
        registerTargetEntityRaycast(entity, callback, opts) {
            registerRaycast(entity, callback, getTargetEntityRaycastDefaultOptions(opts));
        }
    };
}
